src/core/proof.ml

```diff
@@ -9,45 +9,77 @@ open Print
 (** Abstract representation of a goal. *)
 module Goal :
   sig
-    (** Representation of a goal. *)
-    type t
+    (** Representation of a proof type goal. *)
+    type goal_typ = (* TODO have to hide this *)
+      { goal_meta : meta  (* Metavariable needing instantiation.          *)
+      ; goal_hyps : Env.t (* Precomputed scope for a suitable term.       *)
+      ; goal_type : term  (* Precomputed type for a suitable term.        *) }
 
-    (** [of_meta m] create a goal from the metavariable [m]. *)
-    val of_meta : meta -> t
+    (** Representation of a general goal : type, unification *)
+    type t =
+      | Typ of goal_typ (* The usual proof type goal. *)
+      | Unif of constr (* Two terms we'd like equal in some ctxt. *)
+
+    (** Helper functions *)
+    val is_typ  : t -> bool
+    val is_unif : t -> bool
+    val typ : goal_typ -> t
+    val unif : constr  -> t
+    val goal_typ_of : t -> goal_typ
+    val constr_of   : t -> constr
+
+    (** [goal_typ_of_meta m] create a goal from the metavariable [m]. *)
+    val goal_typ_of_meta : meta -> goal_typ
 
     (** [get_meta g] returns the metavariable associated to goal [g]. *)
-    val get_meta : t -> meta
+    val get_meta : goal_typ -> meta
 
     (** [get_type g] returns the environment and expected type of the goal. *)
-    val get_type : t -> Env.t * term
+    val get_type : goal_typ -> Env.t * term
 
     (** [simpl g] simplifies the given goal, evaluating its type to SNF. *)
-    val simpl : t -> t
+    val simpl : goal_typ -> goal_typ
 
     (** Comparison function. *)
-    val compare : t -> t -> int
+    val compare : goal_typ -> goal_typ -> int
   end =
   struct
-    type t =
-      { goal_meta : meta  (* Metavariable needing instantiation.    *)
-      ; goal_hyps : Env.t (* Precomputed scope for a suitable term. *)
-      ; goal_type : term  (* Precomputed type for a suitable term.  *) }
+    type goal_typ =
+      { goal_meta : meta  (* Metavariable needing instantiation.          *)
+      ; goal_hyps : Env.t (* Precomputed scope for a suitable term.       *)
+      ; goal_type : term  (* Precomputed type for a suitable term.        *) }
 
-    let of_meta : meta -> t = fun m ->
+    type t =
+      | Typ of goal_typ (* The usual proof type goal. *)
+      | Unif of constr (* Two terms we'd like equal in ctxt. *)
+
+    let is_typ  = function Typ _ -> true  | Unif _ -> false
+    let is_unif = function Typ _ -> false | Unif _ -> true
+    let typ = function x -> (Typ x)
+    let unif = function x -> (Unif x)
+    let goal_typ_of = function
+      | Typ  gt -> gt
+      | _ -> Console.fatal None "Unif goal instead of a type goal"
```

Comment:
I think that fatal is used for user errors. In case of internal error, we should use assert false or invalid_arg for instance here.

Reply:
Idem below.

Reply:
0df87258 : 

```ocaml
failwith "Internal error : not a type goal"
...
failwith "Internal error : not a type goal"
```